File: blueprints/tasas.py
from flask import Blueprint, jsonify, request
from db import get_db
from auth import admin_required

tasas_bp = Blueprint('tasas', __name__)


@tasas_bp.route('/api/tasas', methods=['GET'])
def get_tasas():
    """Retorna todas las tasas por período o una específica si se pasa year/month."""
    year  = request.args.get('year')
    month = request.args.get('month')
    db    = get_db()

    if year and month:
        row = db.execute(
            'SELECT * FROM tasas_periodo WHERE year=? AND month=?',
            (year, month)
        ).fetchone()
        if not row:
            return jsonify({'error': 'No hay tasas configuradas para este período'}), 404
        return jsonify(dict(row))
    else:
        rows = db.execute('''
            SELECT * FROM tasas_periodo
            ORDER BY year DESC,
            CASE month
                WHEN 'ENE' THEN 1 WHEN 'FEB' THEN 2 WHEN 'MAR' THEN 3 WHEN 'ABR' THEN 4
                WHEN 'MAY' THEN 5 WHEN 'JUN' THEN 6 WHEN 'JUL' THEN 7 WHEN 'AGO' THEN 8
                WHEN 'SEPT' THEN 9 WHEN 'OCT' THEN 10 WHEN 'NOV' THEN 11 WHEN 'DIC' THEN 12
            END DESC
        ''').fetchall()
        return jsonify([dict(r) for r in rows])


@tasas_bp.route('/api/tasas', methods=['POST'])
@admin_required
def save_tasas():
    """
    Guarda o actualiza tasas para un período.
    Body: {year, month, tasa_bcv_inicio, tasa_bcv_fin, tasa_paralela_inicio, tasa_paralela_fin}
    Calcula promedios y factor_diferencial automáticamente.
    """
    data = request.get_json()
    year   = data.get('year')
    month  = data.get('month')
    bcv_ini    = data.get('tasa_bcv_inicio')
    bcv_fin    = data.get('tasa_bcv_fin')
    par_ini    = data.get('tasa_paralela_inicio')
    par_fin    = data.get('tasa_paralela_fin')

    if not all([year, month, bcv_ini is not None, bcv_fin is not None, par_ini is not None, par_fin is not None]):
        return jsonify({'error': 'Faltan parámetros requeridos'}), 400

    if bcv_ini <= 0 or bcv_fin <= 0:
        return jsonify({'error': 'Tasas BCV deben ser mayores a cero'}), 400

    # Calcular promedios
    bcv_prom = (bcv_ini + bcv_fin) / 2
    par_prom = (par_ini + par_fin) / 2
    diferencial = par_prom / bcv_prom

    db = get_db()
    db.execute('''
        INSERT INTO tasas_periodo (year, month, tasa_bcv_inicio, tasa_bcv_fin, tasa_bcv_promedio,
                                   tasa_paralela_inicio, tasa_paralela_fin, tasa_paralela_promedio,
                                   factor_diferencial)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
        ON CONFLICT(year, month) DO UPDATE SET
            tasa_bcv_inicio = excluded.tasa_bcv_inicio,
            tasa_bcv_fin = excluded.tasa_bcv_fin,
            tasa_bcv_promedio = excluded.tasa_bcv_promedio,
            tasa_paralela_inicio = excluded.tasa_paralela_inicio,
            tasa_paralela_fin = excluded.tasa_paralela_fin,
            tasa_paralela_promedio = excluded.tasa_paralela_promedio,
            factor_diferencial = excluded.factor_diferencial
    ''', (year, month, bcv_ini, bcv_fin, bcv_prom, par_ini, par_fin, par_prom, diferencial))
    db.commit()

    return jsonify({
        'ok': True,
        'year': year,
        'month': month,
        'tasa_bcv_promedio': round(bcv_prom, 4),
        'tasa_paralela_promedio': round(par_prom, 4),
        'factor_diferencial': round(diferencial, 4)
    })


@tasas_bp.route('/api/tasas/<int:tasa_id>', methods=['DELETE'])
@admin_required
def delete_tasa(tasa_id):
    db = get_db()
    db.execute('DELETE FROM tasas_periodo WHERE id = ?', (tasa_id,))
    db.commit()
    return jsonify({'ok': True})

File: static/js/tasas.js
// ── CONFIGURACIÓN DE TASAS ────────────────────────────────────────────────
function calcPromedios(){
  const bcvIni=parseFloat(G('tasas-bcv-ini').value)||0;
  const bcvFin=parseFloat(G('tasas-bcv-fin').value)||0;
  const parIni=parseFloat(G('tasas-par-ini').value)||0;
  const parFin=parseFloat(G('tasas-par-fin').value)||0;

  let bcvProm=0, parProm=0;
  if(bcvIni>0&&bcvFin>0){
    bcvProm=(bcvIni+bcvFin)/2;
    G('bcv-promedio').textContent=bcvProm.toFixed(4);
  }else{
    G('bcv-promedio').textContent='—';
  }

  if(parIni>0&&parFin>0){
    parProm=(parIni+parFin)/2;
    G('par-promedio').textContent=parProm.toFixed(4);
  }else{
    G('par-promedio').textContent='—';
  }

  if(bcvProm>0&&parProm>0){
    const dif=(parProm/bcvProm).toFixed(4);
    G('factor-diferencial').textContent=dif;
  }else{
    G('factor-diferencial').textContent='—';
  }
}

async function guardarTasas(){
  if(!isAdmin()){alert('⛔ Acceso denegado: se requiere rol de administrador');return;}
  const year=G('tasas-year').value, month=G('tasas-month').value;
  const bcvIni=parseFloat(G('tasas-bcv-ini').value);
  const bcvFin=parseFloat(G('tasas-bcv-fin').value);
  const parIni=parseFloat(G('tasas-par-ini').value);
  const parFin=parseFloat(G('tasas-par-fin').value);

  if(!bcvIni||!bcvFin||!parIni||!parFin){showA('al-tasas-er','❌ Completa todas las tasas (inicio y fin)');return;}
  if(bcvIni<=0||bcvFin<=0){showA('al-tasas-er','❌ Tasas BCV deben ser mayores a 0');return;}

  const chk=await fetch(`/api/tasas?year=${year}&month=${month}`);
  if(chk.ok){
    const ok=confirm(`Ya existe una tasa guardada para ${month} ${year}. ¿Deseas reemplazarla?`);
    if(!ok)return;
  }

  const r=await fetch('/api/tasas',{method:'POST',headers:{'Content-Type':'application/json'},body:JSON.stringify({
    year,month,
    tasa_bcv_inicio:bcvIni,
    tasa_bcv_fin:bcvFin,
    tasa_paralela_inicio:parIni,
    tasa_paralela_fin:parFin
  })});
  const d=await r.json();
  if(d.ok){
    showA('al-tasas-ok',`✅ Tasas guardadas para ${month} ${year}. Promedio BCV: ${d.tasa_bcv_promedio}, Diferencial: ${d.factor_diferencial}`);
    cargarHistorialTasas();
  }
  else{showA('al-tasas-er','❌ '+d.error);}
}

async function cargarTasas(){
  const year=G('tasas-year').value, month=G('tasas-month').value;
  const r=await fetch(`/api/tasas?year=${year}&month=${month}`);
  if(!r.ok){
    const meses=['ENE','FEB','MAR','ABR','MAY','JUN','JUL','AGO','SEPT','OCT','NOV','DIC'];
    const hoy=new Date();
    const yearActual=String(hoy.getFullYear());
    const mesActualIdx=hoy.getMonth();
    const esPasado=(year<yearActual)||(year===yearActual&&meses.indexOf(month)<mesActualIdx);
    if(esPasado){
      showA('al-tasas-er',`⚠️ No hay tasa registrada para ${month} ${year}`);
    }
    return;
  }
  const d=await r.json();
  G('tasas-bcv-ini').value=d.tasa_bcv_inicio||'';
  G('tasas-bcv-fin').value=d.tasa_bcv_fin||'';
  G('tasas-par-ini').value=d.tasa_paralela_inicio||'';
  G('tasas-par-fin').value=d.tasa_paralela_fin||'';

  calcPromedios();
  showA('al-tasas-ok',`✅ Tasas cargadas: ${month} ${year}`);
}

async function cargarHistorialTasas() {
  try {
    const r = await fetch('/api/tasas');
    const data = await r.json();
    const tbody = G('tasas-historial-tbody');
    G('tasas-historial-loading').style.display = 'none';
    if (!data.length) {
      G('tasas-historial-empty').style.display = 'block';
      G('tasas-historial-wrap').style.display = 'none';
      return;
    }
    G('tasas-historial-empty').style.display = 'none';
    G('tasas-historial-wrap').style.display = 'block';
    tbody.innerHTML = data.map(t => `
      <tr>
        <td>${t.year}</td>
        <td>${t.month}</td>
        <td>${t.tasa_bcv_promedio.toFixed(4)}</td>
        <td>${t.tasa_paralela_promedio.toFixed(4)}</td>
        <td>${t.factor_diferencial.toFixed(4)}</td>
        <td>
          <button class="btnm-sm" onclick="editarTasa('${t.year}','${t.month}')">✏ Editar</button>
          <button class="btnd" onclick="eliminarTasa(${t.id})">🗑 Borrar</button>
        </td>
      </tr>
    `).join('');
  } catch {
    G('tasas-historial-loading').textContent = 'Error al cargar.';
  }
}

async function editarTasa(year, month){
  G('tasas-year').value = year;
  G('tasas-month').value = month;
  await cargarTasas();
  G('tasas-year').scrollIntoView({behavior:'smooth', block:'center'});
}

async function eliminarTasa(id) {
  if (!confirm('¿Eliminar esta tasa? Esta acción no se puede deshacer.')) return;
  const r = await fetch(`/api/tasas/${id}`, { method: 'DELETE' });
  const d = await r.json();
  if (d.ok) cargarHistorialTasas();
}
